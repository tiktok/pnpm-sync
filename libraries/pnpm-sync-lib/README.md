# pnpm-sync-lib

## Introduction

It provides two APIs, ```pnpmSync``` and ```pnpmSyncPrepare```.

## How to install

```
npm i pnpm-sync
```

## How to Use

```
import { pnpmSync, pnpmSyncPrepare } from 'pnpm-sync-lib';
```

