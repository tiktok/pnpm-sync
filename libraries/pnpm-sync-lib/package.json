{
  "name": "pnpm-sync-lib",
  "version": "0.0.1",
  "description": "It provides two APIs, pnpmSync and pnpmSyncPrepare.",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "build": "tsc",
    "start": "tsc -w",
    "prepublishOnly": "npm run build"
  },
  "files": [
    "dist"
  ],
  "author": "chao.guo@bytedance.com",
  "dependencies": {
    "@rushstack/node-core-library": "^3.61.0",
    "@rushstack/package-extractor": "^0.6.12",
    "@microsoft/rush-lib": "~5.110.1",
    "@pnpm/lockfile-file": "~8.1.4",
    "@pnpm/logger": "~5.0.0"
  },
  "devDependencies": {
    "@types/node": "^20.8.10",
    "typescript": "^5.2.2"
  }
}