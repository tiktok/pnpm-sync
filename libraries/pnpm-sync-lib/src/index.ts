export { pnpmSync } from "./pnpmSync";
export { pnpmSyncPrepare } from "./pnpmSyncPrepare";