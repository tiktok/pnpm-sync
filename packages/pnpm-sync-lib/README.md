# pnpm-sync-lib

This is the API library for the [pnpm-sync project](https://www.npmjs.com/package/pnpm-sync).  Please see its documentation for details.
