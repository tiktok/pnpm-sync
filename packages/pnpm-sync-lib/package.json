{
  "name": "pnpm-sync-lib",
  "version": "0.3.3",
  "description": "API library for integrating \"pnpm-sync\" with your toolchain",
  "repository": {
    "type": "git",
    "url": "https://github.com/tiktok/pnpm-sync.git",
    "directory": "packages/pnpm-sync-lib"
  },
  "license": "MIT",
  "main": "lib/index.js",
  "types": "dist/pnpm-sync-lib.d.ts",
  "scripts": {
    "build": "heft build --clean",
    "lint": "eslint src/**",
    "prepublishOnly": "heft build --clean",
    "test": "heft test --clean"
  },
  "dependencies": {
    "@pnpm/dependency-path-pnpm-v10": "npm:@pnpm/dependency-path@^1000.0.9",
    "@pnpm/dependency-path-pnpm-v8": "npm:@pnpm/dependency-path@^2.1.8",
    "@pnpm/dependency-path-pnpm-v9": "npm:@pnpm/dependency-path@^5.1.7",
    "@pnpm/lockfile-types-pnpm-lock-v6": "npm:@pnpm/lockfile-types@^5.1.5",
    "@pnpm/lockfile.types-pnpm-lock-v9": "npm:@pnpm/lockfile.types@^1001.0.8",
    "yaml": "2.4.1"
  },
  "devDependencies": {
    "@rushstack/eslint-config": "^3.6.2",
    "@rushstack/eslint-patch": "^1.7.2",
    "@rushstack/heft": "^0.64.3",
    "@rushstack/heft-node-rig": "^2.4.5",
    "@types/heft-jest": "1.0.6",
    "@types/node": "18.17.15",
    "eslint": "~8.7.0",
    "typescript": "~5.3.3"
  }
}
