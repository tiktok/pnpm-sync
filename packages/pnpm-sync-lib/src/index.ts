/**
 * This package provides an API for automating the actions performed by the
 * {@link https://www.npmjs.com/package/pnpm-sync | pnpm-sync package}.
 *
 * @packageDocumentation
 */

export { pnpmSyncCopy } from "./pnpmSyncCopy";
export { pnpmSyncPrepare } from "./pnpmSyncPrepare";
