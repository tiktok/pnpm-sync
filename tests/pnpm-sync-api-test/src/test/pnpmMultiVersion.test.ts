import path from 'node:path';
import { execa } from 'execa';
import { FileSystem } from '@rushstack/node-core-library';
import { ILogMessageCallbackOptions, pnpmSyncPrepareAsync } from 'pnpm-sync-lib';
import { __dirname, readPnpmLockfile, scrubLog } from './testUtilities.js';

describe('pnpm multi version test', () => {
  it.each(['8', '9', '10'])('pnpm v%s', async (version) => {
    const logs: ILogMessageCallbackOptions[] = [];
    const projectDir = path.join(__dirname, '..', '..', '..', 'test-fixtures', `pnpm-v${version}`);

    // Clean the project directory by removing all untracked files and directories
    await execa({ cwd: projectDir })`git clean -xdf`;
    // Install dependencies using the specified pnpm version via corepack
    await execa({ cwd: projectDir })`corepack pnpm i`;

    await pnpmSyncPrepareAsync({
      lockfilePath: path.join(projectDir, 'pnpm-lock.yaml'),
      dotPnpmFolder: path.join(projectDir, 'node_modules', '.pnpm'),
      ensureFolderAsync: FileSystem.ensureFolderAsync,
      readPnpmLockfile,
      logMessageCallback: (options: ILogMessageCallbackOptions): void => {
        logs.push(options);
      }
    });

    expect(logs.map((x) => scrubLog(x))).toMatchSnapshot();

    const pnpmSyncJSONFile = path.join(
      projectDir,
      'packages',
      'sample-lib1',
      'node_modules',
      '.pnpm-sync.json'
    );
    const pnpmSyncJSON = JSON.parse(FileSystem.readFile(pnpmSyncJSONFile));

    // Verify that the configuration matches the expected snapshot
    expect(pnpmSyncJSON).toMatchSnapshot();

    // Verify that all target folders specified in the configuration exist
    expect(() => {
      for (const targetFolder of pnpmSyncJSON.postbuildInjectedCopy.targetFolders) {
        FileSystem.exists(targetFolder.folderPath);
      }
    }).not.toThrow();
  });
});
