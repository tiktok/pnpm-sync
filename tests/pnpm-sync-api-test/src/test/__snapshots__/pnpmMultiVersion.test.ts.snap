// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pnpm multi version test pnpm v8 1`] = `
Array [
  Object {
    "details": Object {
      "dotPnpmFolder": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v8/node_modules/.pnpm",
      "lockfilePath": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v8/pnpm-lock.yaml",
      "messageIdentifier": "prepare-starting",
    },
    "message": "Starting...",
    "messageKind": "verbose",
  },
  Object {
    "details": Object {
      "messageIdentifier": "prepare-writing-file",
      "pnpmSyncJsonPath": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v8/packages/sample-lib1/node_modules/.pnpm-sync.json",
      "sourceProjectFolder": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v8/packages/sample-lib1",
    },
    "message": "Writing...",
    "messageKind": "verbose",
  },
  Object {
    "details": Object {
      "dotPnpmFolder": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v8/node_modules/.pnpm",
      "executionTimeInMs": "[TIMING]",
      "lockfilePath": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v8/pnpm-lock.yaml",
      "messageIdentifier": "prepare-finishing",
    },
    "message": "Regenerated...",
    "messageKind": "info",
  },
]
`;

exports[`pnpm multi version test pnpm v8 2`] = `
Object {
  "postbuildInjectedCopy": Object {
    "sourceFolder": "..",
    "targetFolders": Array [
      Object {
        "folderPath": "../../../node_modules/.pnpm/file+packages+sample-lib1_@rushstack+eslint-patch@1.11.0_@rushstack+node-core-library@5.13.1__erd7quj6pkdyrimanex2flseii/node_modules/sample-lib1",
      },
      Object {
        "folderPath": "../../../node_modules/.pnpm/file+packages+sample-lib1_@rushstack+eslint-patch@1.11.0_@rushstack+node-core-library@5.13.1__bpmmrb36cawsk5lk5xjymvbs2e/node_modules/sample-lib1",
      },
    ],
  },
  "version": "0.3.3",
}
`;

exports[`pnpm multi version test pnpm v9 1`] = `
Array [
  Object {
    "details": Object {
      "dotPnpmFolder": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v9/node_modules/.pnpm",
      "lockfilePath": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v9/pnpm-lock.yaml",
      "messageIdentifier": "prepare-starting",
    },
    "message": "Starting...",
    "messageKind": "verbose",
  },
  Object {
    "details": Object {
      "messageIdentifier": "prepare-writing-file",
      "pnpmSyncJsonPath": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v9/packages/sample-lib1/node_modules/.pnpm-sync.json",
      "sourceProjectFolder": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v9/packages/sample-lib1",
    },
    "message": "Writing...",
    "messageKind": "verbose",
  },
  Object {
    "details": Object {
      "dotPnpmFolder": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v9/node_modules/.pnpm",
      "executionTimeInMs": "[TIMING]",
      "lockfilePath": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v9/pnpm-lock.yaml",
      "messageIdentifier": "prepare-finishing",
    },
    "message": "Regenerated...",
    "messageKind": "info",
  },
]
`;

exports[`pnpm multi version test pnpm v9 2`] = `
Object {
  "postbuildInjectedCopy": Object {
    "sourceFolder": "..",
    "targetFolders": Array [
      Object {
        "folderPath": "../../../node_modules/.pnpm/sample-lib1@file+packages+sample-lib1_@rushstack+eslint-patch@1.11.0_@rushstack+node-core-lib_jde5jrdqngnzmj5pih5jgr6htu/node_modules/sample-lib1",
      },
    ],
  },
  "version": "0.3.3",
}
`;

exports[`pnpm multi version test pnpm v10 1`] = `
Array [
  Object {
    "details": Object {
      "dotPnpmFolder": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v10/node_modules/.pnpm",
      "lockfilePath": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v10/pnpm-lock.yaml",
      "messageIdentifier": "prepare-starting",
    },
    "message": "Starting...",
    "messageKind": "verbose",
  },
  Object {
    "details": Object {
      "messageIdentifier": "prepare-writing-file",
      "pnpmSyncJsonPath": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v10/packages/sample-lib1/node_modules/.pnpm-sync.json",
      "sourceProjectFolder": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v10/packages/sample-lib1",
    },
    "message": "Writing...",
    "messageKind": "verbose",
  },
  Object {
    "details": Object {
      "dotPnpmFolder": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v10/node_modules/.pnpm",
      "executionTimeInMs": "[TIMING]",
      "lockfilePath": "<root>/pnpm-sync/tests/test-fixtures/pnpm-v10/pnpm-lock.yaml",
      "messageIdentifier": "prepare-finishing",
    },
    "message": "Regenerated...",
    "messageKind": "info",
  },
]
`;

exports[`pnpm multi version test pnpm v10 2`] = `
Object {
  "postbuildInjectedCopy": Object {
    "sourceFolder": "..",
    "targetFolders": Array [
      Object {
        "folderPath": "../../../node_modules/.pnpm/sample-lib1@file+packages+sample-lib1_@rushstack+eslint-patch@1.11.0_@rushstack+node-co_1e4bcf53d5aa94d05e51dfd3fe5a8859/node_modules/sample-lib1",
      },
    ],
  },
  "version": "0.3.3",
}
`;
